daemon: Start DNS proxy before k8s with restored rules

jrajahalme opened the pull request:
Split the endpoint restoration process into more stages so that the saved endpoint DNS history can be used earlier in the bootstrap to start the DNS proxy. This helps ensure that the first regeneration of an Endpoint does not (temporarily) remove the policy keys needed to keep old traffic flowing.

TPROXY rules for the DNS proxy can only be installed couple of seconds later in the bootstrap process. To mitigate this and allow the DNS proxy start serving traffic sooner, it now reads the old TPROXY rule, if available, and tries to re-use the same proxy port number as before the restart. This allows the old datapath config to start feeding traffic to the new proxy instance even before the datapath config can be rewritten.

DNS proxy starts with restored DNS rules based on allowed IP addresses. These rules are removed for each endpoint as soon as the first regeneration completes. Such restored rules should allow DNS requests to be served, but for new DNS resolutions to be added to the Endpoint's policy the restored endpoint's must still have their first regeneration completed.

Fixes: #11004

```release-note
DNS Proxy is started earlier in the Cilium agent bootstrap to make it available to running endpoints sooner.
```

jrajahalme committed:
daemon: Minimize DNS proxy downtime on restart

Start proxy support earlier in the daemon bootstrap, notably before any k8s setup.

Fetch old endpoints earlier so that the DNS history is avalailable
before k8s is set up and move dns proxy initialization earlier in the
bootstrap.

Reuse DNS proxy port from previous run on restart unless overridden by
an explicit Cilium agent option

These changes allow the DNS proxy start serving requests as soon as
the toFQDN policy is received from k8s and avoid any service
disruption prevously possible due to endpoints being regenerated
before the DNS proxy was started.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
endpoint: Update DNSRules on header rewrite

Update DNSRules, if any, before writing headers to capture potentially
changed allowed destination IPs.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

christarazi commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -694,6 +695,28 @@ func (m *IptablesManager) InstallProxyRules(proxyPort uint16, ingress bool, name
 	return m.iptProxyRules("-A", proxyPort, ingress, name)
 }
 
+func GetProxyPort(name string) uint16 {
```
Could you add a godoc for this function?

jrajahalme replied:
Added.

christarazi commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -138,6 +148,73 @@ type portToSelectorAllow map[uint16]cachedSelectorREEntry
+// structure for restored rules that can be used while Cilium agent is restoring endpoints
+type perEPRestored map[uint64]restore.DNSRules
+
+// CheckAllowed checks endpointID, destPort, destID, and name against the rules
```
Nit: this godoc references `CheckAllowed`, but the function name is `checkRestored`

jrajahalme replied:
Copy&paste leftover, fixed.

christarazi commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -138,6 +148,73 @@ type portToSelectorAllow map[uint16]cachedSelectorREEntry
+			ruleIPs[restore.RuleRegex{Regexp: regex}] = IPs
+		}
+		restored[port] = ruleIPs
+	}
```
Are there any performance implications to this? I'm not too familiar with this code, but I assume this nesting is unavoidable for the problem this code is solving. Is it possible to rearrange the data in such a way that we can reduce the nesting? Something to think about as I imagine it won't be trivial to solve now. Could be a followup.

joestringer replied:
Yeah so the interesting question to me is how this differs from the incoming `L7DataMap` form. Practically this would be the full accumulated set of DNS rules acquired to date in the proxy which would be serialized out. Then we're using the selector cache to dump the current evaluations of IPs/identities selected.

For policies where it's just DNS allow-all this should be relatively straightforward (`cs.IsWildcard()` => true) but if you ended up with large numbers of DNS rules it could take some time, any/every time the DNS policy is updated. Would be a good candidate for a metric so we can find out how performance turns out and whether we need to further optimize.

jrajahalme replied:
restore.RuleIPs here is already arranged so that allowed IPs are stored for each regex, while in 'p.allowed' regexes are behind selectors. Here we need to expand the selected IPs for each selector, so the nesting is unavoidable.

In the recommended configuration there would be only one selector (for the kube dns servers) which should only have a small set of IPs.  Wildcard selector is already optimized so that no IPs need to be listed.

Maybe some limits could be applied to guard against pathological policies here.

joestringer replied:
@jrajahalme ^ Please do correct anything I got wrong above :)

christarazi commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -138,6 +148,73 @@ type portToSelectorAllow map[uint16]cachedSelectorREEntry
+	log.Debugf("Restored rules for endpoint %d: %v", endpointID, rules)
+}
+
+func (p *DNSProxy) removeRestoredRules(endpointID uint64) {
```
Could you add a comment stating that this function assume `p` to be locked?

jrajahalme replied:
Done.

christarazi commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -138,6 +148,73 @@ type portToSelectorAllow map[uint16]cachedSelectorREEntry
+	delete(p.restored, endpointID)
+}
+
+func (p *DNSProxy) RemoveRestoredRules(endpointID uint16) {
```
Nit: godoc

jrajahalme replied:
Added.

christarazi commented on pkg/fqdn/dnsproxy/proxy_test.go:
```diff
@@ -524,62 +540,286 @@ func (s *DNSProxyTestSuite) TestFullPathDependence(c *C) {
+	c.Assert(err, Equals, nil, Commentf("Could not unmarshal restored rules from json"))
+	// This comparison does not work, seems like a map element ordering issue,
+	// upon manual inspection the data is the same.
+	// c.Assert(rules, checker.DeepEquals, expected1)
```
This shouldn't be happening...? Source: https://stackoverflow.com/questions/18208394/how-to-test-the-equivalence-of-maps-in-golang

joestringer replied:
+1, unless there's a nested list somewhere they should be equivalent.

jrajahalme replied:
Right. DeepEqual following struct pointer in a map key was not functioning right, though.

I solved this by simplifying the `restore.DNSRules` to eliminate the `*Regexp` as a map key and got rid of this problem :-)

christarazi commented on pkg/fqdn/restore/restore.go:
```diff
@@ -0,0 +1,47 @@
+type IPSet map[string]struct{} // IPs, nil set is wildcard and allows all IPs!
+type RuleRegex struct {
+	*regexp.Regexp
+}
```
Nit: godoc

jrajahalme replied:
Done.

christarazi commented on pkg/fqdn/restore/restore.go:
```diff
@@ -0,0 +1,47 @@
+	"regexp"
+)
+
+// DNS proxy restored rules in a separate package to avoid import loop between pkg/endpoint and pkg/fqd/endpoint
```
You could move this to above the `package restore` line to act like a godoc for the package. 

You could say something like:

```
The restore package provides data structures important to restoring DNS proxy rules. This package serves as a central source for these structures. 
``` 

Something along those lines. :)

jrajahalme replied:
Done, thanks!

christarazi commented on daemon/state.go:
```diff
@@ -115,6 +116,39 @@ func (d *Daemon) validateEndpoint(ep *endpoint.Endpoint) (valid bool, err error)
 }
 
+// fetchOldEndpoints reads the list of existing endpoints previously managed by Cilium when it was
+// last run and associated it with container workloads. This function performs the first step in
```
This mentions "first step in endpoint restoration process" and then there's another function later that mentions second step. What steps are these functions referring to? Are they written down somewhere? It could be helpful to reference that in these comments / godocs.

jrajahalme replied:
This is only really documented in these GoDocs between these two functions you referred to. I'll add a summary and make sure the comments are consistent between each other.

christarazi commented on daemon/state.go:
```diff
@@ -124,16 +158,15 @@ func (d *Daemon) validateEndpoint(ep *endpoint.Endpoint) (valid bool, err error)
-		restored: []*endpoint.Endpoint{},
-		toClean:  []*endpoint.Endpoint{},
-	}
+	defer func() {
+		state.possible = nil
+	}()
```
Why is this necessary?

joestringer replied:
Previously we had a stack variable to track these. Now that the state is shared beyond the scope of one function, if we don't have *something* clean it up then it'll be referenced forever. I think this is logically equivalent to the existing code.

jrajahalme replied:
Not necessary, I used this to make sure I don't dereference it afterwards. Removed.

jrajahalme replied:
Reinstated after reading Joe's comment :-)

joestringer commented on daemon/fqdn.go:
```diff
@@ -30,6 +30,7 @@ import (
 	. "github.com/cilium/cilium/api/v1/server/restapi/policy"
 	"github.com/cilium/cilium/pkg/api"
 	"github.com/cilium/cilium/pkg/controller"
+	"github.com/cilium/cilium/pkg/datapath/iptables"
```
nit: We can avoid the direct iptables dependency by adding the new method to the `Datapath` interface (via `IptablesManager` interface))

```
index 43ed47d5b630..f2ad5088a943 100644
--- a/daemon/fqdn.go
+++ b/daemon/fqdn.go
@@ -30,7 +30,6 @@ import (
        . "github.com/cilium/cilium/api/v1/server/restapi/policy"
        "github.com/cilium/cilium/pkg/api"
        "github.com/cilium/cilium/pkg/controller"
-       "github.com/cilium/cilium/pkg/datapath/iptables"
        "github.com/cilium/cilium/pkg/endpoint"
        "github.com/cilium/cilium/pkg/fqdn"
        "github.com/cilium/cilium/pkg/fqdn/dnsproxy"
@@ -300,7 +299,7 @@ func (d *Daemon) bootstrapFQDN(possibleEndpoints map[uint16]*endpoint.Endpoint,
                port = uint16(option.Config.ToFQDNsProxyPort)
        } else if port == 0 {
                // Try locate old DNS proxy port number from the datapath
-               port = iptables.GetProxyPort(listenerName)
+               port = d.datapath.GetProxyPort(listenerName)
        }
        if err != nil {
                return err
diff --git a/pkg/datapath/iptables/iptables.go b/pkg/datapath/iptables/iptables.go
index 0b4bf9227126..cfb83446b4e6 100644
--- a/pkg/datapath/iptables/iptables.go
+++ b/pkg/datapath/iptables/iptables.go
@@ -695,7 +695,7 @@ func (m *IptablesManager) InstallProxyRules(proxyPort uint16, ingress bool, name
        return m.iptProxyRules("-A", proxyPort, ingress, name)
 }

-func GetProxyPort(name string) uint16 {
+func (m *IptablesManager) GetProxyPort(name string) uint16 {
        prog := "iptables"
        if !option.Config.EnableIPv4 {
                prog = "ip6tables"
diff --git a/pkg/datapath/loader.go b/pkg/datapath/loader.go
index f9c21839be98..ba2aa7ef287b 100644
--- a/pkg/datapath/loader.go
+++ b/pkg/datapath/loader.go
@@ -72,4 +72,8 @@ type IptablesManager interface {
        InstallRules(ifName string) error
        TransientRulesStart(ifName string) error
        TransientRulesEnd(quiet bool)
+
+       // GetProxyPort fetches the existing proxy port configured for the
+       // specified listener. Used early in bootstrap to reopen proxy ports.
+       GetProxyPort(listener string) uint16
 }
```

(This will help subsequently when replacing the underlying implementation)

jrajahalme replied:
Thanks! Done.

christarazi reviewed (changes requested):
Looks good overall. I like the context provided in the PR body, I think it would be helpful to take the relevant pieces of that and put it in the second commit. The first commit has a good amount of context on the "why". 

Some comments, questions to address below.

joestringer commented on daemon/fqdn.go:
```diff
@@ -305,12 +309,25 @@ func (d *Daemon) bootstrapFQDN(restoredEndpoints *endpointRestoreState, preCache
+// updateFQDNrules updates the DNS proxy iptables rules. Must be
+// called after iptables has been initailized, and only after
+// successful bootstrapFQDN().
+func (d *Daemon) updateFQDNrules() error {
```
nit:

```suggestion
// updateDNSrules updates the DNS proxy iptables rules. Must be
// called after iptables has been initailized, and only after
// successful bootstrapFQDN().
func (d *Daemon) updateDNSrules() error {
```

jrajahalme replied:
Done.

jrajahalme replied:
Ended up naming this `updateDNSDatapathRules()` instead to avoid confusion with the restored DNS rules.

joestringer commented on daemon/fqdn.go:
```diff
@@ -136,7 +137,7 @@ func (d *Daemon) updateSelectorCacheFQDNs(ctx context.Context, selectors map[pol
 // default DNS cache. The proxy binds to all interfaces, and uses the
 // configured DNS proxy port (this may be 0 and so OS-assigned).
-func (d *Daemon) bootstrapFQDN(restoredEndpoints *endpointRestoreState, preCachePath string) (err error) {
+func (d *Daemon) bootstrapFQDN(possibleEndpoints map[uint16]*endpoint.Endpoint, preCachePath string) (err error) {
```
Now that `bootstrapFQDN()` is run earlier, do we need to add some kind of channel / delay to the `dns-garbage-collector-job` controller startup so that it doesn't run before all endpoints are restored into the endpointManager?

jrajahalme replied:
At first the garbage collector would not find any endpoints from the endpoint manager, so it would not do anything for yet. And even after the endpoints are exposed, there should be enough slack in the GC to not drop active connections?

joestringer replied:
I see. My main concern was whether the lack of endpoints could clear out global cache entries. But this PR is not loading into the global cache AFAICT and I'm not sure how else it could be prepopulated (maybe `--tofqdns-pre-cache` but I don't think that's often used..?)

joestringer commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -237,7 +318,7 @@ func (proxyStat *ProxyRequestContext) IsTimeout() bool {
 // requesting endpoint. Note that denied requests will not trigger this
 // callback.
-func StartDNSProxy(address string, port uint16, enableDNSCompression bool, lookupEPFunc LookupEndpointIDByIPFunc, lookupSecIDFunc LookupSecIDByIPFunc, notifyFunc NotifyOnDNSMsgFunc) (*DNSProxy, error) {
+func StartDNSProxy(address string, port uint16, enableDNSCompression bool, lookupEPFunc LookupEndpointIDByIPFunc, lookupSecIDFunc LookupSecIDByIPFunc, lookupIPsFunc LookupIPsBySecIDFunc, notifyFunc NotifyOnDNSMsgFunc) (*DNSProxy, error) {
```
nit: By this point it makes sense for us to introduce some kind of 'owner' interface type to encapsulate the various functions we pass in here. Though maybe we should make that part of a longer-term cleanup effort.

jrajahalme replied:
Leaving this as a follow-up work item.

joestringer commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -138,6 +148,73 @@ type portToSelectorAllow map[uint16]cachedSelectorREEntry
+	return restored
+}
+
+// RestoreRules is used in the beginning of endpoint restoration to
+// install rules saved before the restart to be used before the endpoint
+// is regenerated.
+func (p *DNSProxy) RestoreRules(endpointID uint16, rules restore.DNSRules) {
```
Is there any particular reason to have this separate from `StartDNSProxy()`?

jrajahalme replied:
Only a practical one, as `StartDNSProxy()` handles the whole proxy, while restored rules are available for specific endpoints. I originally patterned this based on DNSHistory, but that is not fed to DNSProxy.

joestringer replied:
I see. My question was thinking from two perspectives: One is if we have all data at the beginning, create the object with all available data and secondly, by calling in multiple times, we end up grabbing locks several times for each endpoint. I don't think either are particularly pressing issues.

joestringer replied:
@jrajahalme actually, does this increase the window of time where the DNS proxy is rejecting connections? Is it possible to set these up before opening the proxy socket?

joestringer commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -694,6 +695,28 @@ func (m *IptablesManager) InstallProxyRules(proxyPort uint16, ingress bool, name
+
+	res, err := runProgCombinedOutput(prog, []string{"-t", "mangle", "-n", "-L", ciliumPreMangleChain}, false)
+	if err != nil {
+		return 0
```
This is supposed to be triggered after iptables initialization, right? So it should always succeed? By implication, perhaps we should add a warning here just so that if any of those constraints are not met, then we'll at least have a log telling us that we're silently failing to restore the proxy port due to the lack of chain.

joestringer replied:
I guess the flip side is that we have an "info" log describing which port we select when we *do* succeed, so maybe this is not necessary.

jrajahalme replied:
This is only supposed to work if the **previous** run installed rules, so this should work before any iptables initialization on the current run.

joestringer commented on daemon/state.go:
```diff
@@ -269,7 +288,10 @@ func (d *Daemon) regenerateRestoredEndpoints(state *endpointRestoreState) (resto
-			if err := ep.RegenerateAfterRestore(); err != nil {
+			err := ep.RegenerateAfterRestore()
+			// Remove restored rules after the initial regeneration
+			proxy.DefaultDNSProxy.RemoveRestoredRules(ep.ID)
```
Regeneration might not have been successful in this case, I guess by unconditionally running this here we're stating that  the endpoint is in a bad state, so it's fine to kill DNS connectivity now?

The alternative would be to only clear out the restored rules upon success, which would be more like "coast and keep doing what you do until we know we've successfully replaced the datapath".

jrajahalme replied:
Right, I thought about this but flipped the logic in my mind :-( Fixed.

joestringer commented on pkg/ipcache/ipcache.go:
```diff
@@ -443,13 +443,16 @@ func (ipc *IPCache) LookupByPrefix(IP string) (Identity, bool) {
-func (ipc *IPCache) LookupByIdentity(id identity.NumericIdentity) (map[string]struct{}, bool) {
+// security identity ID, or nil if the entry does not exist.
+// Can't return the internal map as it may be modifeid at any time when the
+// lock is not held, so return a slice of strings instead
```
2 nits on comment: `modifeid` -> `modified`; this is more of an implementation detail so could go inside the function rather than here which ends up in godoc.

jrajahalme replied:
Fixed and moved, thanks!

joestringer commented on pkg/ipcache/ipcache.go:
```diff
@@ -443,13 +443,16 @@ func (ipc *IPCache) LookupByPrefix(IP string) (Identity, bool) {
+	for ip := range ipc.identityToIPCache[id] {
+		ips = append(ips, ip)
+	}
+	return ips
```
nit for memory allocations:

```suggestion
	ips = make([]string, 0, len(ipc.identityToIPCache[id])
	for ip := range ipc.identityToIPCache[id] {
		ips = append(ips, ip)
	}
	return ips
```

jrajahalme replied:
Added.

joestringer commented on pkg/proxy/proxy.go:
```diff
@@ -254,6 +254,18 @@ func findProxyPort(name string) *ProxyPort {
 
+// AckProxyPort() marks the proxy at 'pp' as successfully created and creates or updates the datapath rules
+// accordingly.
+func (p *Proxy) AckProxyPort(name string) error {
```
nit: The one caller always does `GetProxyPort()` prior to this call to fetch the filter name, then we end up locking & looking up `pp` in that function, then returning, and then calling into here. Did you consider having the parameters here match `GetProxyPort()` to hide that complexity from the caller?

jrajahalme replied:
I did not, but that is a good point, call site becomes trivial :-) Done.

joestringer commented on pkg/fqdn/dnsproxy/proxy_test.go:
```diff
@@ -524,62 +540,286 @@ func (s *DNSProxyTestSuite) TestFullPathDependence(c *C) {
+	_, exists = s.proxy.restored[epID1]
+	c.Assert(exists, Equals, true)
+
+	// Same tests with 2 (WORLD) dstID to make sure it is not used, but with correct destionation IP
```
```suggestion
	// Same tests with 2 (WORLD) dstID to make sure it is not used, but with correct destination IP
```

jrajahalme replied:
Fixed, thanks!

joestringer commented on pkg/fqdn/dnsproxy/proxy_test.go:
```diff
@@ -524,62 +540,286 @@ func (s *DNSProxyTestSuite) TestFullPathDependence(c *C) {
+	// Case 1 | EPID1 | dstIP1 |   53 | www.ubuntu.com | Allowed
+	allowed, err = s.proxy.CheckAllowed(epID1, 53, 2, dstIP1, "www.ubuntu.com")
+	c.Assert(err, Equals, nil, Commentf("Error when checking allowed"))
+	c.Assert(allowed, Equals, true, Commentf("request was rejected when it should be allowed"))
```
nit: The comments in this section could benefit from mentioning that the allow is expected from the "restore of DNS rules", just to differentiate from earlier failures.

jrajahalme replied:
Missed this sorry, will address in the next round if needed.

jrajahalme replied:
Done.

joestringer commented on pkg/fqdn/dnsproxy/proxy_test.go:
```diff
@@ -524,62 +540,286 @@ func (s *DNSProxyTestSuite) TestFullPathDependence(c *C) {
+	_, exists = s.proxy.restored[epID1]
+	c.Assert(exists, Equals, false)
+
+	// Before restore: everything is dropped
```
What's the difference between this section and the section from lines 646-708?

jrajahalme replied:
before the earlier block all normal rules were removed. Here we removed the current restored rules. If that removal did not work, we might not notice json marshal/unmarshal problems in the following tests.

jrajahalme replied:
I changed the comments of these two test blocks to add context.

joestringer reviewed (changes requested):
Most of the below are minor nits that could be followed up afterwards.  PTAL & ping me to tighten the feedback loop on anything not marked "nit", and we can file an issue for any nits you don't address here.

Looking back through my main question is on starting the DNS GC controller, but I did have some other questions too.

jrajahalme committed:
endpoint: Remove restored DNS rules

Remove restored DNS rules after a successful regeneration, and also at
endpoint delete to cover endpoints that were never regenerated.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

joestringer commented on pkg/endpoint/bpf.go:
```diff
@@ -144,6 +144,18 @@ func (e *Endpoint) writeHeaderfile(prefix string) error {
+	// DNSRules (== nil) we don't need to update here, as in that case there are no allowed
+	// destinations either.
+	if e.DNSRules != nil {
+		e.OnDNSPolicyUpdateLocked(e.owner.GetDNSRules(e.ID))
```
I thought we were deliberately only updating this when the policy changes to reduce load here :-)

This runs every 5s. Are we happy enough with the performance in pathological cases to do this frequently?

jrajahalme replied:
The set of selected IPs can change any time. The cost of figuring out if any of the selectors now selects new IDs and whether the IPs they map to are still the same is roughly the same as recreating the structure anyway?

jrajahalme replied:
Maybe we should have some limit for the max number of rules and/or IPs we process?

joestringer replied:
> The set of selected IPs can change any time

Ah right, I keep thinking of this as storing the policy layer but we must realize the selectors to provide meaningful data on restore. :+1: 

> The cost of figuring out if any of the selectors now selects new IDs and whether the IPs they map to are still the same is roughly the same as recreating the structure anyway?

:+1: 

> Maybe we should have some limit for the max number of rules and/or IPs we process?

That seems somewhat reasonable if we're trying to bound the time, but without knowing the cost it's hard to know what sensible values would be here or whether we would simply introduce additional problems by adding a limiter. 

Are there simple high-signal outputs we could get here that would point when if this is *actually* expensive at runtime, not just based on my wild speculation? I'm thinking again something like metrics or a very rare warning message. Seems like we expect this to be pretty cheap despite the `GetRules()` nesting that Chris pointed out in the thread above (https://github.com/cilium/cilium/pull/12718#discussion_r463154010).

joestringer replied:
Resolved in the latest patch.

joestringer commented on daemon/state.go:
```diff
@@ -309,6 +309,9 @@ func (d *Daemon) regenerateRestoredEndpoints(state *endpointRestoreState) (resto
 	var endpointCleanupCompleted sync.WaitGroup
 	for _, ep := range state.toClean {
+		// Remove restored rules of cleaned endpoint
+		proxy.DefaultDNSProxy.RemoveRestoredRules(ep.ID)
```
Should this be handled instead via either `Endpoint.Delete()` or `Endpoint.leaveLocked()`?

jrajahalme replied:
Good point. That would cover the failed regeneration case as well, when the endpoint ends up being deleted eventually!

joestringer replied:
Allegedly yes :-) I know we had a bug around this in the past where one path didn't properly clean up the endpoint but I *think* we converged the implementation towards actually calling the delete functions in each path. Maybe @christarazi recalls the details on what I'm alluding to here? I think it was related to metrics for endpoints on restore.

christarazi replied:
Yes that's correct. The commit in question is here: https://github.com/cilium/cilium/pull/11884/commits/07a2171e64afebaecd0c2fc82cb8dae99b9d1723

Basically, if there was an attempt to create an endpoint with invalid data, the above commit now deletes that endpoint, rather than leaving it. All paths are covered with that commit, so I believe if there's a new angle that ends up with endpoints left around, then we should delete them.

jrajahalme replied:
WIll also have to move the removal after initial regeneration to happen from the endpoint after any successful regeneration to cover the case where the initial regeneration fails.

christarazi replied:
Will this be done as a followup?

jrajahalme committed:
dnsproxy: Use restored rules during restart

Store current DNS rules with the Endpoint and use them in the DNS
proxy during initial regeneration of the restrored endpoints.

DNS proxy starts with restored DNS rules based on allowed IP
addresses. These rules are removed for each endpoint as soon as the
first regeneration completes. Such restored rules should allow DNS
requests to be served, but for new DNS resolutions to be added to the
Endpoint's policy the restored endpoint's must still have their first
regeneration completed.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

christarazi commented on daemon/state.go:
```diff
@@ -115,25 +117,64 @@ func (d *Daemon) validateEndpoint(ep *endpoint.Endpoint) (valid bool, err error)
+//
+// 1. fetchOldEndpoints(): Unmarshal old endpoints
+//    - used to start DNS proxy with restored DNS history and rules
+// 2. restoreOldEndpoints(): validate andpoint data after k8s has been configured
```
Nit: 
```suggestion
// 2. restoreOldEndpoints(): validate endpoint data after k8s has been configured
```

jrajahalme replied:
Fixed.

jrajahalme committed:
fqdn: Limit the max processing in GetRules()

Normally the number of DNS proxy rules should be very small. To guard
against pathological cases, limit the number of IPs processed to 1000
per port.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

joestringer commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -175,18 +175,28 @@ func (p *DNSProxy) GetRules(endpointID uint16) restore.DNSRules {
 						IPs[ip] = struct{}{}
+						count++
+						if count > 1000 {
+							log.Warningf("Too many DNS rules or destinations for a port, skipping the rest")
```
Can we include some more details here for debugging, such as endpoint ID? I guess port may not necessarily be all that interesting as it's typically 53 but we could include that too. maybe `cs`?

jrajahalme replied:
Sure, a moment...

joestringer replied:
Oh and `Warning()`, not `Warningf()`.

jrajahalme replied:
Done.

christarazi commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -175,18 +175,32 @@ func (p *DNSProxy) GetRules(endpointID uint16) restore.DNSRules {
 					for _, ip := range p.LookupIPsBySecID(nid) {
 						IPs[ip] = struct{}{}
+						count++
+						if count > 1000 {
```
Nice. Just curious how was 1000 chosen? Might be helpful to document why in the commit msg or similar for our future selves in case this ever comes back to bite us. (Unlikely, but worth spending the bit of extra effort :) )

jrajahalme replied:
The concern here is that there would be a pathological policy that allows huge number of destinations (but not all, as the wildcard case is already optimized) and that would cause too much load as this thing is done every 5 seconds on each endpoint. Could have picked also 100, but felt that maybe some large loadbalanced DNS service might come too close to it. So basically 1000 felt like large enough to never cause trouble, while at the same time making sure we do not burn too much CPU doing this. WIll leave this like this for now having the good green CI run down there :-)

joestringer reviewed (approved):
I think this is much better than where we were before. LGTM. There might be minor followups but I'm not seeing any of them as crucial as part of this PR.
